src/foam/nanos/auth/CapabilityAuthService.js

```diff
@@ -71,32 +142,61 @@ foam.CLASS({
             )
           );
 
-          DAO userCapabilityJunctionDAO = (DAO) x.get("userCapabilityJunctionDAO");
+          if ( cap != null && cap.getEnabled() ) {
 
-          if ( cap != null && 
-               userCapabilityJunctionDAO.find(
-                AND(
-                  capabilityScope,
-                  EQ(UserCapabilityJunction.TARGET_ID, cap.getId()),
-                  EQ(UserCapabilityJunction.STATUS, CapabilityJunctionStatus.GRANTED)
-             )) != null ) 
-            return true;
-
-          List<UserCapabilityJunction> userCapabilityJunctions = ((ArraySink) user.getCapabilities(x).getJunctionDAO()
-            .where(capabilityScope)
-            .select(new ArraySink())).getArray();
-
-          for ( UserCapabilityJunction ucJunction : userCapabilityJunctions ) {
-            Capability capability = (Capability) capabilityDAO.find(ucJunction.getTargetId());
-            if ( capability.implies(x, permission) ) return true;
+            if ( userCapabilityJunctionDAO.find(
+              AND(
+                capabilityScope,
+                EQ(UserCapabilityJunction.TARGET_ID, cap.getId()),
+                EQ(UserCapabilityJunction.STATUS, CapabilityJunctionStatus.GRANTED)
+              )) != null ) result = true;
+            // if the user has the permission, store this in the cache and return the result
+            // otherwise, move on to the 2nd part of the check
+            if ( result ) {
+              ((Map<String, Boolean>) getCache()).put(key, result);
+              return result;
+            }
           }
+
+          // 2. check if the user has a capability that grants the permission
+          ProxySink proxy = new ProxySink(x, new LimitedSink(x, 1, 0, new ArraySink())) {
+            int count = 0;
+            @Override
+            public void put(Object o, Detachable sub) {
+              UserCapabilityJunction ucj = (UserCapabilityJunction) ((UserCapabilityJunction) o).deepClone();
```

Comment:
Why do you need to deepClone() the UCJ?

Reply:
I got these 2 comments via email but they are not showing up on the page
> What is the ProxySink for?
> Couldn't you just write a Predicate and use where() instead?
ProxySink is used to override put and only put the ucjs where the capability's implies method returns true for the permission. Do we have a way to do this with predicate?

> Why do you need to deepClone() the UCJ?
Sorry I can't remember why either.. but I don't think this is needed, will remove